### PR TITLE
tutorial: proposal for kdd 2024

### PR DESCRIPTION
**Pull request by ncclementi**
Still needs work, and reviews, I left comments in the document. Please include your feedback. 

xref: https://github.com/ibis-project/talks/issues/15

### CODE REVIEW

File: kdd-2024/ibis-tutorial-KDD2024.md [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+- I'm not sure who to put here, I will need designated responsible people for some 
+  Ibis ML showcase and Ibis streaming (Chloe?)
+
+
+# Corresponding tutor with her/his email address
+
+- Naty Clementi (I'll add my email upon submission)
+
+# Tutorial outline. Please provide as much detail as possible.
```
Comment (ncclementi) [inline kdd-2024/ibis-tutorial-KDD2024.md:127]: This is the part that needs more work and help to be cohesive and  include streaming and ML.

File: kdd-2024/ibis-tutorial-KDD2024.md [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+pandas’ dataframe interface, requiring a rewrite of your analysis code.
+
+This is where Ibis comes in. Ibis is a pure-Python open-source library, originally 
+created by Wes McKinney creator of pandas, that provides a dataframe interface to
+many popular databases and analytics tools (DuckDB, Polars, Snowflake, Spark, etc). This 
+lets users analyze data using the same consistent API, regardless of which backend 
+they’re using, and without ever having to learn SQL. This tutorial introduces 
+Ibis, emphasizing its role in simplifying data manipulation across diverse platforms.
+
+Additionally, attendees will explore Ibis' advancements in data streaming technology. The
```
Comment (chloeh13q) [inline kdd-2024/ibis-tutorial-KDD2024.md:38]: I would rephrase "Ibis' advancements in data streaming technology" to perhaps "Ibis' advancements in stream-batch unification"

File: kdd-2024/ibis-tutorial-KDD2024.md [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+This is where Ibis comes in. Ibis is a pure-Python open-source library, originally 
+created by Wes McKinney creator of pandas, that provides a dataframe interface to
+many popular databases and analytics tools (DuckDB, Polars, Snowflake, Spark, etc). This 
+lets users analyze data using the same consistent API, regardless of which backend 
+they’re using, and without ever having to learn SQL. This tutorial introduces 
+Ibis, emphasizing its role in simplifying data manipulation across diverse platforms.
+
+Additionally, attendees will explore Ibis' advancements in data streaming technology. The 
+Ibis-flink backend, inspired by Apache Flink SQL, combines stream and batch into a single
```
Comment (chloeh13q) [inline kdd-2024/ibis-tutorial-KDD2024.md:39]: Can take out "inspired by Apache Flink SQL" if we're running out of space

File: kdd-2024/ibis-tutorial-KDD2024.md [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+many popular databases and analytics tools (DuckDB, Polars, Snowflake, Spark, etc). This 
+lets users analyze data using the same consistent API, regardless of which backend 
+they’re using, and without ever having to learn SQL. This tutorial introduces 
+Ibis, emphasizing its role in simplifying data manipulation across diverse platforms.
+
+Additionally, attendees will explore Ibis' advancements in data streaming technology. The 
+Ibis-flink backend, inspired by Apache Flink SQL, combines stream and batch into a single 
+framework, addressing common barriers faced by data scientists venturing into 
+streaming analytics. By leveraging Ibis unified Python dataframe API, we facilitate 
+seamless transitions between batch and streaming workflows, reducing operational
```
Comment (chloeh13q) [inline kdd-2024/ibis-tutorial-KDD2024.md:42]: I like this! :)

File: kdd-2024/ibis-tutorial-KDD2024.md [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+
+## 2:30 - UDFs and sql passthrough (I THINK THIS SHOULD BE GONE GIVEN THE AUDIENCE AND TIME)
+
+Demonstrate using UDFs to add custom operations.
+
+Explain and demonstrate various "escape hatches" when you really need to use SQL directly.
+
+# 3:00 - Coffee Break (5 minutes) + Q&A in the room
+
+## HELP HERE MAYBE AN EXAMPLE THAT CAN SHOW STREAMING AND ML, OR TWO? IT'LL HAVE TO BE SMALL
```
Comment (chloeh13q) [inline kdd-2024/ibis-tutorial-KDD2024.md:187]: +1 on things being disconnected. 

I think [something like this](https://ibis-project.org/posts/unified-stream-batch/) is a good example; generally the demos that we do for ibis streaming show some type of window aggregation in Flink (because window aggregations play an important role, especially in a lot of the ML-powered use cases that we were looking at), then execute the same window aggregation in DuckDB/another batch engine. 

My concern would be: would we need to spend time introducing Flink/streaming? I think it may be hard for the audience to understand why we're doing this / why this is important if they have little understanding of the problem space (the message may be blurry - they're all just backends, now a streaming backend, so what?). 

Perhaps 5 minutes talking about the problem / streaming and ~10 minutes walking through an example?

File: kdd-2024/ibis-tutorial-KDD2024.md [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+
+Additionally, attendees will explore Ibis' advancements in data streaming technology. The 
+Ibis-flink backend, inspired by Apache Flink SQL, combines stream and batch into a single 
+framework, addressing common barriers faced by data scientists venturing into 
+streaming analytics. By leveraging Ibis unified Python dataframe API, we facilitate 
+seamless transitions between batch and streaming workflows, reducing operational
+complexities and enabling real-time machine learning applications.
+
+Last but not least, this tutorial will also introduce Ibis ML, an evolving library 
+for building machine learning pipelines using Ibis. Where these pipelines can then
```
Comment (lostmygithubaccount) [inline kdd-2024/ibis-tutorial-KDD2024.md:46]: Ibis ML -> IbisML

Perhaps in place of the second sentence, something like:

"IbisML brings the portability of Ibis to ML pipelines -- preprocess your data on any of the 20+ backends Ibis supports, then efficiently handoff to popular training libraries like scikit-learn, XGBoost, or PyTorch."

File: kdd-2024/ibis-tutorial-KDD2024.md [hunk 7]
```diff
@@ -0,0 +1,273 @@
+# Descriptive Title 
+
+We need a title here, in the past we had this options
+
+- Scipy and PyCon 2024: Intro to Ibis: blazing fast analytics with DuckDB, Polars, Snowflake, and more,
```
Comment (lostmygithubaccount) [inline kdd-2024/ibis-tutorial-KDD2024.md:5]: How about: "Ibis: portable and blazing fast analytics with DuckDB, Polars, Snowflake, and more..."

**Review by lostmygithubaccount — COMMENTED** [review]
thanks for taking this on!

File: kdd-2024/ibis-tutorial-KDD2024.md (hunk 5 above)
Comment (gforsyth) [reply kdd-2024/ibis-tutorial-KDD2024.md:187]: @chloeh13q @deepyaman do either of you have any experience adding data generators like those used in the flink examples?

If it isn't too cumbersome, my thinking is we can have it fake incoming IMDB ratings, so we start with a batched version of the problem, using that to introduce Ibis as we build up expressions, then we can pivot to an example that involves a stream of incoming reviews (obviously contrived, but I think it would be reasonably representative)

File: kdd-2024/ibis-tutorial-KDD2024.md [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+The target audience for this hands-on tutorial is data scientists, data and machine
+learning engineers, researchers, practitioners, and industry professionals who are 
+already familiar with data analysis and data manipulation tools in Python. 
+
+They may have encountered scalability challenges with some libraries like pandas 
+and are interested in exploring alternative solutions to improve performance and
+efficiency.
+
+Additionally, individuals interested in data streaming technology and machine 
+learning applications would find value in learning about Ibis' advancements in
```
Comment (gforsyth) [inline kdd-2024/ibis-tutorial-KDD2024.md:66]:
```suggestion
learning applications will find value in learning about Ibis' advancements in 
```

File: kdd-2024/ibis-tutorial-KDD2024.md [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+They may have encountered scalability challenges with some libraries like pandas 
+and are interested in exploring alternative solutions to improve performance and
+efficiency.
+
+Additionally, individuals interested in data streaming technology and machine 
+learning applications would find value in learning about Ibis' advancements in 
+these areas. 
+
+This is a hands-on tutorial, with numerous examples to get your hands on 
+experience. Participants should ideally have some experience using Python and
```
Comment (gforsyth) [inline kdd-2024/ibis-tutorial-KDD2024.md:70]:
```suggestion
This is a hands-on tutorial, with numerous examples. Participants should ideally have some experience using Python and
```

File: kdd-2024/ibis-tutorial-KDD2024.md [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+Get attendees up and running in a GitHub Codespace or on their laptops. A bit of
+motivation about the kinds of problems where Ibis can help, and a general survey 
+of attendees to find out what their existing pain points and experiences are.
+
+## 0:15 - Introduction to Ibis basics
+
+**Slides**
+- What is Ibis?: Ibis as part of a composable data ecosystem.
+- Why Ibis?: present the problems that Ibis tackles and the benefits of adopting it.
+- Who Is Ibis for?: shown the versatility of profiles that can benefit from Ibis (data engineers, data analysts, and data scientists, any data practitioner) and how.
```
Comment (gforsyth) [inline kdd-2024/ibis-tutorial-KDD2024.md:146]:
```suggestion
- Who Is Ibis for?: show how data engineers, data analysts, data scientists, and any data practitioner can benefit from Ibis
```

File: kdd-2024/ibis-tutorial-KDD2024.md [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+
+## 1:10 - In-memory tables, joins, and data analysis (need to adapt this for audience)
+
+Building on the previous notebook, we'll explore how to join in-memory data (from 
+a pandas DataFrame, Python dictionary, or PyArrow Table) with existing tables in 
+a local database and continue analysis on the join result.
+
+We'll touch on the Ibis deferred operator for specifying predicates in chained joins, 
+and demonstrate read_parquet and other read_* methods for loading local data into
+existing databases. Then we'll continue with a series of hands-on exercises,
```
Comment (gforsyth) [inline kdd-2024/ibis-tutorial-KDD2024.md:164]:
```suggestion
We'll explore chained joins, and demonstrate read_parquet and other read_* methods for loading local data into
existing databases. Then we'll continue with a series of hands-on exercises, 
```

we no longer _need_ the underscore for chained joins, so we can introduce it elsewhere.

File: kdd-2024/ibis-tutorial-KDD2024.md [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+## 2:10 - Selectors 
+Continuing on from joins, we'll introduce selectors as a means of quickly 
+renaming and cleaning datasets, a powerful feature stolen inspired by dplyr.
+
+## 2:30 - UDFs and sql passthrough (I THINK THIS SHOULD BE GONE GIVEN THE AUDIENCE AND TIME)
+
+Demonstrate using UDFs to add custom operations.
+
+Explain and demonstrate various "escape hatches" when you really need to use SQL directly.
+
```
Comment (gforsyth) [inline kdd-2024/ibis-tutorial-KDD2024.md:184]:
```suggestion
```

agreed on cutting this section

File: kdd-2024/ibis-tutorial-KDD2024.md [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+execution, or with bursting to a cloud database (or other hosted database).
+
+## 1:50 - Coffee Break (5 minutes) + Q&A in the room
+
+HERE IS WHERE WE NEED TO THINK HOW TO TIE THE PREVIOUS CONTENT WITH STREAMING AND ML 
+WHAT WE HAVE IN THE SCIPY TUTORIAL IS 
+
+## 2:10 - Selectors 
+Continuing on from joins, we'll introduce selectors as a means of quickly 
+renaming and cleaning datasets, a powerful feature stolen inspired by dplyr.
```
Comment (gforsyth) [inline kdd-2024/ibis-tutorial-KDD2024.md:177]:
```suggestion
renaming and cleaning datasets, a powerful feature ~stolen from~ inspired by dplyr.
```

**Review by gforsyth — COMMENTED** [review]
going to take another pass, but some small comments / changes

I think we will probably drop the entire PyPI section and use that section for focusing on streaming and IbisML stuff

File: kdd-2024/ibis-tutorial-KDD2024.md (hunk 3 above)
Comment (chloeh13q) [reply kdd-2024/ibis-tutorial-KDD2024.md:39]: I think both - Flink SQL is Flink’s effort in stream-batch unification, and it’s been rather successful, which inspired us to push the boundaries further

File: kdd-2024/ibis-tutorial-KDD2024.md (hunk 5 above)
Comment (chloeh13q) [reply kdd-2024/ibis-tutorial-KDD2024.md:187]: Yes we have done this before. Flink has a datagen connector; alternatively in the demos we spin up a Kafka topic and use the python-kafka client to write data into it

File: kdd-2024/ibis-tutorial-KDD2024.md (hunk 3 above)
Comment (ncclementi) [reply kdd-2024/ibis-tutorial-KDD2024.md:39]: to avoid confusion I'll remove the "inspired by Apache Flink SQL,"

File: kdd-2024/ibis-tutorial-KDD2024.md [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+
+I tried to adapt what we had for SciPy and PyCon, and include something on Streaming 
+and ML. Keep in mind that streaming and ML will be a small part of this tutorial, per 
+Ian's suggestion 20%, please make suggestions with this in mind. 
+
+**(I believe we are at 270 words now)**
+
+Tabular data analysis, a staple in data science, often encounters scalability 
+challenges with pandas. In response, modern analytical databases (like DuckDB) 
+offer significant performance boosts but typically demand SQL expertise. Many of
```
[the remark this answers is not in the document]
Comment (ncclementi) [reply kdd-2024/ibis-tutorial-KDD2024.md:27]: @lostmygithubaccount I'mm not sure I follow the suggestion, would it be this edit? or something different? 
 
```suggestion
offer significant performance boosts but typically demand SQL preference abd expertise. Many of 
```

File: kdd-2024/ibis-tutorial-KDD2024.md (hunk 7 above)
Comment (ncclementi) [reply kdd-2024/ibis-tutorial-KDD2024.md:5]: I'm worried that we won't capture the DS and ML crowd by using the word "analytics" but idk.

File: kdd-2024/ibis-tutorial-KDD2024.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (lostmygithubaccount) [reply kdd-2024/ibis-tutorial-KDD2024.md:27]: yes that is what I meant

**Comment by chloeh13q** [comment]
### Chloe He

**Bio**

Chloe has a background in data science and started working on streaming systems when she was a Founding Engineer at Claypot AI, a startup tackling challenges in real-time machine learning. She led the infrastructure development of an open-source real-time feature engineering platform and worked on the translating and optimizing streaming workloads that served low-latency use cases. Later, she brought her streaming expertise to Voltron Data, where she now leads the development of Ibis streaming.

**Expertise:**

Chloe has presented talks and workshops across a number of conferences, such as DataConnect, NVIDIA GTC, and Google DevFest. She also taught lectures for the Machine Learning System Design course at Stanford and a crash course on recommendation systems.

File: kdd-2024/ibis-tutorial-KDD2024.md [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+**Bio**
+Chloe has a background in data science and started working on streaming systems when she was a Founding Engineer at Claypot AI, a startup tackling challenges in real-time machine learning. She led the infrastructure development of an open-source real-time feature engineering platform and worked on the translating and optimizing streaming workloads that served low-latency use cases. Later, she brought her streaming expertise to Voltron Data, where she now leads the development of Ibis streaming.
+
+**Expertise:**
+Chloe has presented talks and workshops across a number of conferences, such as DataConnect, NVIDIA GTC, and Google DevFest. She also taught lectures for the Machine Learning System Design course at Stanford and a crash course on recommendation systems.
+
+
+2. List of contributors, i.e., tutors who will only help prepare the tutorial material
+
+- Cody Peterson
```
Comment (ncclementi) [inline kdd-2024/ibis-tutorial-KDD2024.md:110]: @lostmygithubaccount and @deepyaman  

I'm thinking of you two to help contributing with whatever we will do for ML, happy to remove you if you won't be able to help or think the other one got it.

File: kdd-2024/ibis-tutorial-KDD2024.md [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+
+We'll build on our previous batch analysis example and turn what we've learned
+towards a similar problem, but now with a real-time and constantly updating data
+source. 
+
+With hands-on exercises, we'll explore how to set up rudimentary anomaly
+detection, and demonstrate how to window and group streaming data-sources before
+redirecting them to downstream sinks.
+
+## IbisML
```
Comment (ncclementi) [inline kdd-2024/ibis-tutorial-KDD2024.md:164]: @lostmygithubaccount or @chloeh13q can you help writing one sentence of what we will aim to do here?

**Comment by ncclementi** [comment]
Also team, wee need to add a section of 

> A brief discussion of the potential societal impacts of your tutorial

I have no clue what to say here. @lostmygithubaccount any ideas?

**Comment by lostmygithubaccount** [comment]
IMO saving developer time wasted. how much time has been spent on database migrations and code migrations because everyone who builds anything also builds their own API for it

File: kdd-2024/ibis-tutorial-KDD2024.md (hunk 16 above)
Comment (chloeh13q) [reply kdd-2024/ibis-tutorial-KDD2024.md:164]: "We will demonstrate how to define machine learning features, create data transformation pipelines, and exeucte them across backends using IbisML. We will then use these features to train an ML model!"

Comment (gforsyth) [reply kdd-2024/ibis-tutorial-KDD2024.md:164]: I think the above is good, but we might want to mention something about how you can also do this without the usual memory constraints

Comment (ncclementi) [reply kdd-2024/ibis-tutorial-KDD2024.md:187]: I added what Chloe mention in the docs, would you mind taking a crack at modifying the sentence to "mention something about how you can also do this without the usual memory constraints" 

I'm not fully sure, what we want to say here, by usual memory constraints . Is it like "running out of memory during and xgboost training" kind of thing?

File: kdd-2024/ibis-tutorial-KDD2024.md [hunk 17]
```diff
@@ -0,0 +1,286 @@
...
+and leave participants feeling motivated and satisfied.
+
+- **One Up, One Down Feedback:** At the end of the day, learners will provide
+alternating positive and negative feedback points about the day's session. This 
+structured feedback encourages participants to express their honest opinions, 
+helping instructors to address concerns effectively.
+
+
+# A brief discussion of the potential societal impacts of your tutorial
+
```
Comment (ncclementi) [inline kdd-2024/ibis-tutorial-KDD2024.md:272]: I took a crack at this, let me know if I over did it. Happy to cut some things over.

File: kdd-2024/ibis-tutorial-KDD2024.md (hunk 16 above)
Comment (gforsyth) [reply kdd-2024/ibis-tutorial-KDD2024.md:187]: Yeah, I think this is less important to cover in the outline, so I would leave it as is.

File: kdd-2024/ibis-tutorial-KDD2024.md [hunk 18]
```diff
@@ -0,0 +1,286 @@
+# Descriptive Title 
+
+- "Ibis: The Data Scientist's Powerhouse for Blazing-Fast Analytics and Machine Learning"
```
Comment (zhenzhongxu) [inline kdd-2024/ibis-tutorial-KDD2024.md:3]: nit: Powerhouse and Blazing-fast are both buzzwords, I'd suggest giving a more direct title on why Ibis is the powerhouse and why it's blazing-fast on large scale data

Comment (ncclementi) [reply kdd-2024/ibis-tutorial-KDD2024.md:3]: Sure, Do you have a suggestion? We were trying to get a title that targets DS and ML.

Comment (zhenzhongxu) [reply kdd-2024/ibis-tutorial-KDD2024.md:3]: how about something like: "Unified Machine Learning & Analytics workflows with Ibis: bridging batch and streaming from local notebook to production scale"

Comment (lostmygithubaccount) [reply kdd-2024/ibis-tutorial-KDD2024.md:3]: I would change "Powerhouse" to "Toolkit" -- I don't think the focus of this talk as-is is on bridging batch to local though. we should probably focus on our strengths today (Analytics) and the next-level things (ML, Streaming)

perhaps:

"Ibis: The Data Scientist's Toolkit for Blazing-Fast Analytics, Machine Learning, and Stream Processing"

Comment (gforsyth) [reply kdd-2024/ibis-tutorial-KDD2024.md:3]: boom.  :shipit:

Comment (ncclementi) [inline kdd-2024/ibis-tutorial-KDD2024.md:3]:
```suggestion
- "Ibis: The Data Scientist's Toolkit for Blazing-Fast Analytics, Machine Learning, and Stream Processing"
```

File: kdd-2024/ibis-tutorial-KDD2024.md [hunk 19]
```diff
@@ -0,0 +1,286 @@
...
+
+With hands-on exercises, we'll explore how to set up rudimentary anomaly
+detection, and demonstrate how to window and group streaming data-sources before
+redirecting them to downstream sinks.
+
+## IbisML
+
+We will demonstrate how to define machine learning features, create data
+transformation pipelines, and execute them across backends using IbisML. We will
+then use these features to train an ML model
```
Comment (ncclementi) [inline kdd-2024/ibis-tutorial-KDD2024.md:191]:
```suggestion
then use these features to train an ML model.
```

File: kdd-2024/ibis-tutorial-KDD2024.md [hunk 20]
```diff
@@ -0,0 +1,286 @@
...
+framework, addressing common barriers faced by data scientists venturing into 
+streaming analytics. By leveraging Ibis' unified Python dataframe API, we facilitate 
+seamless transitions between batch and streaming workflows, reducing operational
+complexities and enabling real-time machine learning applications.
+
+Last but not least, this tutorial will also introduce IbisML, an evolving library 
+for building machine learning pipelines using Ibis. IbisML brings the portability
+of Ibis to ML pipelines -- preprocess your data on any of the 20+ backends that 
+Ibis supports, then efficiently handoff to popular training libraries like
+scikit-learn, XGBoost, or PyTorch
```
Comment (ncclementi) [inline kdd-2024/ibis-tutorial-KDD2024.md:34]:
```suggestion
scikit-learn, XGBoost, or PyTorch.
```

**Review by gforsyth — APPROVED** [review]
Thanks for leading this, @ncclementi !
